pkg/prometheus/operator.go

```diff
@@ -1310,6 +1325,94 @@ func (c *Operator) loadBasicAuthSecrets(
 
 }
 
+func (c *Operator) loadBearerTokensFromSecrets(mons map[string]*monitoringv1.ServiceMonitor) (map[string]BearerToken, error) {
+	tokens := map[string]BearerToken{}
+	nsSecretCache := make(map[string]*v1.Secret)
+
+	for _, mon := range mons {
+		for i, ep := range mon.Spec.Endpoints {
+			if ep.BearerTokenSecret.Name == "" {
+				continue
+			}
+
+			sClient := c.kclient.CoreV1().Secrets(mon.Namespace)
+			token, err := getCredFromSecret(
+				sClient,
+				ep.BearerTokenSecret,
+				"bearertoken",
+				mon.Namespace+"/"+ep.BearerTokenSecret.Name,
+				nsSecretCache,
+			)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"failed to extract endpoint bearertoken for servicemonitor %v from secret %v in namespace %v",
+					mon.Name, ep.BearerTokenSecret.Name, mon.Namespace,
+				)
+			}
+
+			tokens[fmt.Sprintf("serviceMonitor/%s/%s/%d", mon.Namespace, mon.Name, i)] = BearerToken(token)
+		}
+	}
+
+	return tokens, nil
+}
+
+func (c *Operator) loadTLSAssetsFromSecrets(mons map[string]*monitoringv1.ServiceMonitor) (map[string]TLSAsset, error) {
+	assets := map[string]TLSAsset{}
+	nsSecretCache := make(map[string]*v1.Secret)
+
+	for _, mon := range mons {
+		for _, ep := range mon.Spec.Endpoints {
+			if ep.TLSConfig == nil {
+				continue
+			}
+
+			prefix := mon.Namespace + "/"
+			secretSelectors := map[string]*v1.SecretKeySelector{}
+			if ep.TLSConfig.CASecret != nil {
+				secretSelectors[prefix+ep.TLSConfig.CASecret.Name+"/"+ep.TLSConfig.CASecret.Key] = ep.TLSConfig.CASecret
+			}
+			if ep.TLSConfig.CertSecret != nil {
+
+				secretSelectors[prefix+ep.TLSConfig.CertSecret.Name+"/"+ep.TLSConfig.CertSecret.Key] = ep.TLSConfig.CertSecret
+			}
+			if ep.TLSConfig.KeySecret != nil {
+				secretSelectors[prefix+ep.TLSConfig.KeySecret.Name+"/"+ep.TLSConfig.KeySecret.Key] = ep.TLSConfig.KeySecret
+			}
+
+			for key, selector := range secretSelectors {
+				sClient := c.kclient.CoreV1().Secrets(mon.Namespace)
+				asset, err := getCredFromSecret(
+					sClient,
+					*selector,
+					"tls config",
+					// TODO: Is the cache key really necessary? Can't this be part of `getCredFromSecret`?
+					key,
+					nsSecretCache,
+				)
+				if err != nil {
+					return nil, fmt.Errorf(
+						"failed to extract endpoint tls asset for servicemonitor %v from secret %v and key %v in namespace %v",
+						mon.Name, selector.Name, selector.Key, mon.Namespace,
+					)
+				}
+
+				// TODO: Namespacing via underscores seems rather hacky.
+				assets[fmt.Sprintf(
+					"%v_%v_%v",
```

Comment:
This might be a good candidate for a const too :blush:

Reply:
Right. Maybe we can even combine this with `promcfg.go/addTLStoYaml` to keep the two in sync. Do you have a good alternative to the hacky `_` namespacing? That might turn into a security concern given that the following scenarios both return the same path:

- namespace `foo_xy`, secret name `bar`, key name `baz` => `foo_xy_bar_baz`
- namespace `foo`, secret name `xy_bar`, key name `baz` => `foo_xy_bar_baz`

Reply:
As the [docs](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#creating-a-new-namespace) suggest:

> Note that the name of your namespace must be a DNS compatible label.

Same applies for secrets names. 

Given this limitation I think it should be OK to leave the underscore as delimiter (given there cannot be ns/secret objects containing

```
$ kubectl create ns some_ns
The Namespace "some_ns" is invalid: metadata.name: Invalid value: "some_ns": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')

$ kubectl create secret generic some_secret --from-literal=foo=bar
The Secret "some_secret" is invalid: metadata.name: Invalid value: "some_secret": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

this means that underscores cannot be found on regular ns/secrets names hence I think it should be safe for usage in this case.